lib/scholar/cluster/k_means.ex

```diff
@@ -0,0 +1,337 @@
+
+    centroid_mask = Nx.indexed_put(centroid_mask, indices_mask, Nx.broadcast(0, {num_runs}))
+
+    {_, _, _, _, final_centroids} =
+      while {idx = 1, centroid_mask, x, inertia, centroids},
+            Nx.less(idx, num_clusters) do
+        {_inertia_for_centroids, min_inertia} =
+          calculate_inertia(x, centroids, num_clusters, num_runs)
+
+        {new_centroid, centroid_idx} = find_new_centroid(min_inertia, x, num_clusters, num_runs)
+
+        indices_centroids =
+          Nx.stack(
+            [
+              Nx.flatten(Nx.iota({num_runs, num_features}, axis: 0)),
+              Nx.broadcast(idx, {num_runs * num_features}),
+              Nx.tile(Nx.iota({num_features}), [num_runs])
+            ],
+            axis: -1
+          )
+
+        centroids = Nx.indexed_put(centroids, indices_centroids, Nx.flatten(new_centroid))
+
+        indices_mask = Nx.stack([Nx.iota({num_runs}), centroid_idx], axis: -1)
+
+        centroid_mask = Nx.indexed_put(centroid_mask, indices_mask, Nx.broadcast(0, {num_runs}))
+        {idx + 1, centroid_mask, x, inertia, centroids}
+      end
+
+    final_centroids
+  end
+
+  # Function checks validity of the provided data
+
+  deftransformp verify(x, opts) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            "expected x to have shape {n_samples, n_features}, got tensor with shape: #{inspect(Nx.shape(x))}"
+    end
+
+    {num_samples, _num_features} = Nx.shape(x)
+
+    unless opts[:num_clusters] do
+      raise ArgumentError,
+            "missing option :num_clusters"
+    end
+
+    unless is_integer(opts[:num_clusters]) and opts[:num_clusters] > 0 and
+             opts[:num_clusters] <= num_samples do
+      raise ArgumentError,
+            "expected :num_clusters to to be a positive integer in range 1 to #{inspect(num_samples)}, got: #{inspect(opts[:num_clusters])}"
+    end
+
+    unless opts[:init] in [:random, :k_means_plus_plus] do
+      raise ArgumentError,
+            "expected :init to be either :random or :k_means_plus_plus, got: #{inspect(opts[:init])}"
+    end
+
+    unless is_integer(opts[:max_iterations]) and opts[:max_iterations] > 0 do
+      raise ArgumentError,
+            "expected :max_iterations to be a positive integer, got: #{inspect(opts[:max_iterations])}"
+    end
+
+    unless is_integer(opts[:num_runs]) and opts[:num_runs] > 0 do
+      raise ArgumentError,
+            "expected :num_runs to be a positive integer, got: #{inspect(opts[:num_runs])}"
+    end
+
+    unless is_number(opts[:tol]) and opts[:tol] >= 0 do
+      raise ArgumentError,
+            "expected :tol to be a non-negative number, got: #{inspect(opts[:tol])}"
+    end
+
+    unless is_integer(opts[:random_state]) or is_nil(opts[:random_state]) do
+      raise ArgumentError,
+            "expected :random_state to be an integer or nil, got: #{inspect(opts[:random_state])}"
+    end
+
+    if opts[:random_state] != nil do
+      :rand.seed(:exsss, opts[:random_state])
```

Comment:
This will only impact the BinaryBackend and this happens at defn compile time, which we don't want. I think we need to wait for https://github.com/elixir-nx/nx/pull/846 to do this sort of things.

As for tests that rely on this, maybe we could use data points that guarantee convergence, regardless the initialization. And/or have some property tests if there are any we can reasonably have.

Reply:
The problem will be only with the labels parameter. Depending on the initialization of centroids the resultant labels might be i. ex. [1, 0, 1, 0], but also [0, 1, 0, 1].

Reply:
You can normalize he classes for comparison.

```elixir
defmodule Scholar.Utils do
  import Nx.Defn

  @doc """
  Returns labels that unambiguously identify the given grouping.

  This function effectively renumbers the classes, such that the
  first occurrences of each classes are in increasing order.
  
  ## Examples

      iex> Scholar.Utils.normalize_labels(Nx.tensor([1, 0, 1, 0]), num_classes: 2)
      #Nx.Tensor<
        s64[4]
        [0, 1, 0, 1]
      >

      iex> Scholar.Utils.normalize_labels(Nx.tensor([1, 2, 2, 0, 1, 2, 0]), num_classes: 3)
      #Nx.Tensor<
        s64[7]
        [0, 1, 1, 2, 0, 1, 2]
      >

  """
  defn normalize_labels(labels, opts \\ []) do
    opts = keyword!(opts, [:num_classes])

    num_classes = opts[:num_classes]

    classes = Nx.iota({num_classes})

    class_min_idx =
      Nx.equal(
        Nx.new_axis(classes, 1),
        Nx.new_axis(labels, 0)
      )
      |> Nx.argmax(axis: 1)

    mapping = to_consecutive(class_min_idx)

    Nx.take(mapping, labels)
  end

  # This is effectively a double argsort, just optimised
  defnp to_consecutive(x) do
    x = Nx.argsort(x)
    Nx.indexed_add(Nx.broadcast(0, x), Nx.new_axis(x, -1), Nx.iota(x))
  end
end
```